sui_types/src/message_headers.rs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
+pub enum TraceTag {
+    LatencyProbe,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Header {
+    TraceId(u64),
+    TraceTag(TraceTag),
+}
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct SerializedHeaders {
+    // Zero or more Header instances. Only the first
+    // occurrence of a variant is read, subsequent ones
+    // are discarded.
+    pub headers: Vec<Header>,
+}
+
+#[derive(Default)]
+pub struct Headers {
```

Comment:
1) Please add RustDoc everywhere
2) I can't tell what the purpose of this struct is.... it seems we would always want a `Vec<Header>` everywhere right?    I also don't understand why we need a separate `SerializedHeaders`.   Couldn't we use a single structure that serializes correctly but is still generic?

Reply:
My concern with having two structures is that there is a very large overhead.  For each message we would have to: 
1. Deserialize the envelope, which creates SerializedHeaders
2. The SerializedHeaders then creates a Headers, which has to interpret every header

The ideal for performance would be a BTreeMap with the enum tag number as the key, for quick lookup.  (Well actually a HashMap but we aren't allowed to have those)

Reply:
Why can't we use hash tables? determinism?

Re performance, we either have to process every header and put it into some fast data structure, or else have O(n) access to headers later. I don't see a way around that.

Most implementations (e.g. HTTP, gRPC), use the former approach.

Putting things into a map of any kind is going to be quite awkward, since headers may not all have the same type, therefore each value in the map will have to be in an enum. That's why I did it this way.

If we need to switch to a map later, we can do it internally and cope with the ugliness inside `Headers`. But for now, this way is faster: as long as we know the set of all possible headers we may as well put them into a struct and get constant time access and convenient typing.